pkg/clients/database/sql.go

```diff
@@ -0,0 +1,612 @@
+		TenantID:     c.TenantID,
+		AADEndpoint:  c.ActiveDirectoryEndpointURL,
+		Resource:     c.ResourceManagerEndpointURL,
+	}
+	a, err := cfg.Authorizer()
+	if err != nil {
+		return nil, errors.Wrapf(err, "cannot create Azure authorizer from credentials config")
+	}
+	client.Authorizer = a
+	if err := client.AddToUserAgent(azure.UserAgent); err != nil {
+		return nil, errors.Wrap(err, "cannot add to Azure client user agent")
+	}
+
+	return client, nil
+}
+
+// NewPostgreSQLVirtualNetworkRuleParameters returns an Azure VirtualNetworkRule object from a virtual network spec
+func NewPostgreSQLVirtualNetworkRuleParameters(v *azuredbv1alpha3.PostgreSQLServerVirtualNetworkRule) postgresql.VirtualNetworkRule {
+	return postgresql.VirtualNetworkRule{
+		Name: azure.ToStringPtr(v.Spec.Name),
+		VirtualNetworkRuleProperties: &postgresql.VirtualNetworkRuleProperties{
+			VirtualNetworkSubnetID:           azure.ToStringPtr(v.Spec.VirtualNetworkRuleProperties.VirtualNetworkSubnetID),
+			IgnoreMissingVnetServiceEndpoint: azure.ToBoolPtr(v.Spec.VirtualNetworkRuleProperties.IgnoreMissingVnetServiceEndpoint, azure.FieldRequired),
+		},
+	}
+}
+
+// PostgreSQLServerVirtualNetworkRuleNeedsUpdate determines if a virtual network rule needs to be updated
+func PostgreSQLServerVirtualNetworkRuleNeedsUpdate(kube *azuredbv1alpha3.PostgreSQLServerVirtualNetworkRule, az postgresql.VirtualNetworkRule) bool {
+	up := NewPostgreSQLVirtualNetworkRuleParameters(kube)
+
+	switch {
+	case !reflect.DeepEqual(up.VirtualNetworkRuleProperties.VirtualNetworkSubnetID, az.VirtualNetworkRuleProperties.VirtualNetworkSubnetID):
+		return true
+	case !reflect.DeepEqual(up.VirtualNetworkRuleProperties.IgnoreMissingVnetServiceEndpoint, az.VirtualNetworkRuleProperties.IgnoreMissingVnetServiceEndpoint):
+		return true
+	}
+
+	return false
+}
+
+// UpdatePostgreSQLVirtualNetworkRuleStatusFromAzure updates the status related to the external
+// Azure PostgreSQLVirtualNetworkRule in the VirtualNetworkStatus
+func UpdatePostgreSQLVirtualNetworkRuleStatusFromAzure(v *azuredbv1alpha3.PostgreSQLServerVirtualNetworkRule, az postgresql.VirtualNetworkRule) {
+	v.Status.State = string(az.VirtualNetworkRuleProperties.State)
+	v.Status.ID = azure.ToString(az.ID)
+	v.Status.Type = azure.ToString(az.Type)
+}
+
+// Helper functions
+// NOTE: These helper functions work for both MySQL and PostreSQL, but we cast everything to the MySQL types because
+// the generated Azure clients for MySQL and PostgreSQL are exactly the same content, just a different package. See:
+// https://github.com/Azure/azure-sdk-for-go/blob/master/services/mysql/mgmt/2017-12-01/mysql/models.go
+// https://github.com/Azure/azure-sdk-for-go/blob/master/services/postgresql/mgmt/2017-12-01/postgresql/models.go
+
+// The name must match the specification of the SKU, so, we don't allow user
+// to specify an arbitrary name.
+
+// ToMySQLSKU returns the name of the MySQL Server SKU, which is tier + family + cores, e.g. B_Gen4_1, GP_Gen5_8.
```

Comment:
Does this return just the name? The comment reads a little like it's returning a (string?) name, but it seems to be returning a pointer to struct.